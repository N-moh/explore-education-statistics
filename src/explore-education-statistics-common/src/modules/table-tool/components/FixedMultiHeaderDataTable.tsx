import { FullTableMeta } from '@common/modules/full-table/types/fullTable';
import Footnote from '@common/modules/table-tool/components/Footnote';
import React, { forwardRef, ReactNode, Ref, useEffect, useRef } from 'react';
import styles from './FixedMultiHeaderDataTable.module.scss';
import MultiHeaderTable, { RowHeaderType } from './MultiHeaderTable';

const mobileWidth = 1024;

interface Props {
  caption: ReactNode;
  captionId?: string;
  innerRef?: Ref<HTMLElement>;
  columnHeaders: string[][];
  rowHeaders: RowHeaderType[][];
  rowHeaderIsGroup?: boolean[];
  rows: string[][];
  footnotes?: FullTableMeta['footnotes'];
}

const FixedMultiHeaderDataTable = forwardRef<HTMLElement, Props>(
  (props, ref) => {
    const { caption, captionId = 'dataTableCaption', footnotes = [] } = props;

    const containerRef = useRef<HTMLDivElement>(null);
    const mainTableRef = useRef<HTMLTableElement>(null);

    useEffect(() => {
      const listener = () => {
        if (
          containerRef.current &&
          mainTableRef.current &&
          window.innerWidth <= mobileWidth
        ) {
          mainTableRef.current
            .querySelectorAll<HTMLTableHeaderCellElement>('thead td')
            .forEach(el => {
              // eslint-disable-next-line no-param-reassign
              el.style.transform = '';
            });

          mainTableRef.current
            .querySelectorAll<HTMLTableHeaderCellElement>('thead th')
            .forEach(el => {
              // eslint-disable-next-line no-param-reassign
              el.style.transform = '';
            });

          mainTableRef.current
            .querySelectorAll<HTMLTableHeaderCellElement>('tbody th')
            .forEach(el => {
              // eslint-disable-next-line no-param-reassign
              el.style.transform = '';
            });
        }
      };

      window.addEventListener('resize', listener);

      return () => window.removeEventListener('resize', listener);
    }, []);

    return (
      <figure className={styles.figure} ref={ref}>
        <figcaption>{caption}</figcaption>

        <div
          className={styles.container}
          ref={containerRef}
          role="region"
          tabIndex={-1}
          onScroll={event => {
            const { scrollLeft, scrollTop } = event.currentTarget;

            if (mainTableRef.current && window.innerWidth >= mobileWidth) {
              mainTableRef.current
                .querySelectorAll<HTMLTableHeaderCellElement>('thead td')
                .forEach(el => {
                  // eslint-disable-next-line no-param-reassign
                  el.style.transform = `translate(${scrollLeft}px, ${scrollTop}px)`;
                });

              mainTableRef.current
                .querySelectorAll<HTMLTableHeaderCellElement>('thead th')
                .forEach(el => {
                  // eslint-disable-next-line no-param-reassign
                  el.style.transform = `translate(0, ${scrollTop}px)`;
                });

              mainTableRef.current
                .querySelectorAll<HTMLTableHeaderCellElement>('tbody th')
                .forEach(el => {
                  // eslint-disable-next-line no-param-reassign
                  el.style.transform = `translate(${scrollLeft}px, 0)`;
                });
            }
          }}
        >
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            ref={mainTableRef}
          />
        </div>
        {footnotes.length > 0 && (
          <>
            <h3>Footnotes</h3>
            <Footnote content={footnotes} />
          </>
        )}
      </figure>
    );
  },
);

FixedMultiHeaderDataTable.displayName = 'FixedMultiHeaderDataTable';

export default FixedMultiHeaderDataTable;
