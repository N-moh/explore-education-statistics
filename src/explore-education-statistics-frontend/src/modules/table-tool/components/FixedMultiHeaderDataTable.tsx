import throttle from 'lodash/throttle';
import React, { forwardRef, Ref, useEffect, useRef } from 'react';
import DataTableKeys from './DataTableKeys';
import styles from './FixedMultiHeaderDataTable.module.scss';
import MultiHeaderTable from './MultiHeaderTable';

interface Props {
  caption: string;
  captionId?: string;
  innerRef?: Ref<HTMLElement>;
  columnHeaders: string[][];
  rowHeaders: string[][];
  rows: string[][];
}

const FixedMultiHeaderDataTable = forwardRef<HTMLElement, Props>(
  (props, ref) => {
    const { caption, captionId = 'dataTableCaption' } = props;

    const mainTableRef = useRef<HTMLTableElement>(null);
    const headerTableRef = useRef<HTMLTableElement>(null);
    const columnTableRef = useRef<HTMLTableElement>(null);
    const intersectionTableRef = useRef<HTMLTableElement>(null);

    const setStickyElementSizes = throttle(() => {
      if (mainTableRef.current) {
        const mainTableEl = mainTableRef.current;

        const cloneCellHeightWidth = (
          selector: string,
          tableEl: HTMLTableElement,
        ) => {
          const tableCells = tableEl.querySelectorAll<HTMLTableCellElement>(
            selector,
          );

          mainTableEl
            .querySelectorAll<HTMLTableCellElement>(selector)
            .forEach((el, index) => {
              tableCells[index].style.height = `${el.offsetHeight}px`;
              tableCells[index].style.width = `${el.offsetWidth}px`;
            });
        };

        if (headerTableRef.current) {
          headerTableRef.current.style.width = `${mainTableEl.offsetWidth}px`;

          cloneCellHeightWidth('thead th', headerTableRef.current);
        }

        if (columnTableRef.current) {
          cloneCellHeightWidth(
            'thead tr:first-child td:first-child',
            columnTableRef.current,
          );
          cloneCellHeightWidth('tbody th', columnTableRef.current);
        }

        if (intersectionTableRef.current) {
          cloneCellHeightWidth(
            'thead tr:first-child td:first-child',
            intersectionTableRef.current,
          );
        }
      }
    }, 200);

    useEffect(() => {
      setTimeout(() => {
        setStickyElementSizes();

        window.addEventListener('resize', setStickyElementSizes);
      }, 200);

      return () => {
        window.removeEventListener('resize', setStickyElementSizes);
      };
    });

    return (
      <figure className={styles.figure} ref={ref}>
        <figcaption>
          <strong id={captionId}>{caption}</strong>

          <DataTableKeys />
        </figcaption>

        <div
          className={styles.container}
          role="region"
          tabIndex={-1}
          onScroll={event => {
            const { scrollLeft, scrollTop } = event.currentTarget;

            if (headerTableRef.current) {
              headerTableRef.current.style.top = `${scrollTop}px`;
            }

            if (columnTableRef.current) {
              columnTableRef.current.style.left = `${scrollLeft}px`;
            }

            if (intersectionTableRef.current) {
              intersectionTableRef.current.style.top = `${scrollTop}px`;
              intersectionTableRef.current.style.left = `${scrollLeft}px`;
            }
          }}
        >
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            className={styles.intersectionTable}
            ref={intersectionTableRef}
            ariaHidden
            isStickyColumn
            isStickyHeader
          />
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            className={styles.columnTable}
            ref={columnTableRef}
            ariaHidden
            isStickyColumn
          />
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            className={styles.headerTable}
            ref={headerTableRef}
            ariaHidden
            isStickyHeader
          />

          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            ref={mainTableRef}
          />
        </div>
      </figure>
    );
  },
);

FixedMultiHeaderDataTable.displayName = 'FixedHeaderGroupedDataTable';

export default FixedMultiHeaderDataTable;
