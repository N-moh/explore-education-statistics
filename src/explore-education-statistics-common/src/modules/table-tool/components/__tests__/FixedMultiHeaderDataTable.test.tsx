import React from 'react';
import { render } from 'react-testing-library';
import FixedMultiHeaderDataTable from '../FixedMultiHeaderDataTable';

describe('FixedMultiHeaderDataTable', () => {
  test('renders underlying table', () => {
    const { container } = render(
      <FixedMultiHeaderDataTable
        caption="Test table"
        columnHeaders={[
          ['Col group A', 'Col group B'],
          ['Col group C', 'Col group D'],
        ]}
        rowHeaders={[['Row group A', 'Row group B']]}
        rows={[
          ['1', '2', '3', '4'],
          ['5', '6', '7', '8'],
        ]}
      />,
    );

    const table = container.querySelector('table') as HTMLElement;

    expect(table.querySelectorAll('thead tr')).toHaveLength(2);
    expect(table.querySelectorAll('thead tr:nth-child(1) th')).toHaveLength(2);
    expect(table.querySelectorAll('thead tr:nth-child(2) th')).toHaveLength(4);

    expect(table.querySelectorAll('tbody tr')).toHaveLength(2);
    expect(table.querySelectorAll('tbody tr:nth-child(1) th')).toHaveLength(1);
    expect(table.querySelectorAll('tbody tr:nth-child(2) th')).toHaveLength(1);

    expect(table).toMatchSnapshot();
  });
});
