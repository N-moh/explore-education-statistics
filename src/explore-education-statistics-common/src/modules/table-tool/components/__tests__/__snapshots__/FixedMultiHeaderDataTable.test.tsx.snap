// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedMultiHeaderDataTable renders underlying table 1`] = `
<table
  aria-labelledby="dataTableCaption"
  class="govuk-table table"
>
  <thead>
    <tr>
      <td
        class="borderBottom borderRight"
        colspan="1"
        rowspan="2"
      />
      <th
        class="govuk-table__header--center"
        colspan="2"
        scope="colgroup"
      >
        Col group A
      </th>
      <th
        class="govuk-table__header--center"
        colspan="2"
        scope="colgroup"
      >
        Col group B
      </th>
    </tr>
    <tr>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group C
      </th>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group D
      </th>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group C
      </th>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group D
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th
        class="govuk-table__cell--numeric"
        rowspan="1"
        scope="row"
      >
        Row group A
      </th>
      <td
        class="govuk-table__cell--numeric borderBottom"
      >
        1
      </td>
      <td
        class="govuk-table__cell--numeric borderRight borderBottom"
      >
        2
      </td>
      <td
        class="govuk-table__cell--numeric borderBottom"
      >
        3
      </td>
      <td
        class="govuk-table__cell--numeric borderRight borderBottom"
      >
        4
      </td>
    </tr>
    <tr>
      <th
        class="govuk-table__cell--numeric"
        rowspan="1"
        scope="row"
      >
        Row group B
      </th>
      <td
        class="govuk-table__cell--numeric borderBottom"
      >
        5
      </td>
      <td
        class="govuk-table__cell--numeric borderRight borderBottom"
      >
        6
      </td>
      <td
        class="govuk-table__cell--numeric borderBottom"
      >
        7
      </td>
      <td
        class="govuk-table__cell--numeric borderRight borderBottom"
      >
        8
      </td>
    </tr>
  </tbody>
</table>
`;
